client/pkg/transport/listener.go

```diff
@@ -444,6 +448,17 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
 			return cert.VerifyHostname(info.AllowedHostname) == nil
 		}
 	}
+
+	var definedRestrictions int
+	for _, restriction := range []int{len(info.AllowedCNs), len(info.AllowedHostnames), len(info.AllowedURIs)} {
+		if restriction > 0 {
+			definedRestrictions++
+			if definedRestrictions > 1 {
+				return nil, errors.New("exactly one of AllowedCNs, AllowedHostnames, or AllowedURIs can be defined")
+			}
+		}
+	}
+
```

Comment:
Since https://github.com/etcd-io/etcd/pull/18160 introduced more flags than the original patch, we would need to remove these lines https://github.com/DataDog/etcd/blob/282099b9a89a76d0dc2281b6c9563c3acc335e53/client/pkg/transport/listener.go#L426-L437

And change the for loop to iterate over something like `[]int{len(info.AllowedCN), len(info.AllowedCNs), len(info.AllowedHostname), len(info.AllowedHostnames), len(info.AllowedURIs)}`

Reply:
In 881159b, I modified the for loop:

Running through the logic now:

If we change the for loop to be: `[]int{len(info.AllowedCN), len(info.AllowedCNs), len(info.AllowedHostname), len(info.AllowedHostnames), len(info.AllowedURIs)}`

then we're only allowed one of:
- `AllowedCN`
- `AllowedCNs`
- `AllowedHostname`
- `AllowedHostnames`
- `AllowedURIs`
- 
I think the mutual exclusion is also encapsulated in this b/c you can't have:

- `AllowedCN` and `AllowedCNs`
- `AllowedHostname` and `AllowedHostnames`
- `AllowedCNs` and `AllowedHostnames`
- `AllowedCNs` and `AllowedURIs`
- 
there are others included, but I'm not going to list them all.

I wonder though if we modify the for loop above, can we still keep those lines? They'd be superfluous but it'd still keep us _more_ consistent with upstream. This is my first time doing a cherry-pick so I'm not sure what the best practice is.

Reply:
I think the current upstream implementation is hardly scalable as we add new fields, and it already misses some edge cases

The goal here is to only allow one of the flags and supersede `verifyCertificate` accordingly
As such, we should check all pairs for mutual exclusion, which in combinatorics is in square of the number of flags, e.g. there are 6 pairs with CN and Hostname, and there will be 10 pairs with CN and Hostname

The current upstream implementation only has 4 cases and not 6, and indeed it does not prevent having `AllowedCN` and `AllowedHostnames` together, or `AllowedCNs` and `AllowedHostname` together

In the end, I believe there is an opportunity to fix this bug upstream
We should write our patch with this fix in mind by removing the manual check for manual exclusions and only rely on the for loop